browser/brave_vpn/dns/brave_vpn_dns_observer_service.cc

```diff
@@ -0,0 +1,277 @@
+  return secure || is_valid_automatic_mode || is_system_doh_enabled;
+}
+
+bool BraveVpnDnsObserverService::ShouldAllowExternalChanges() const {
+  if (allow_changes_for_testing_.has_value())
+    return allow_changes_for_testing_.value();
+
+  return (chrome::ShowQuestionMessageBoxSync(
+              GetAnchorBrowserWindow(),
+              l10n_util::GetStringUTF16(IDS_PRODUCT_NAME),
+              l10n_util::GetStringUTF16(IDS_BRAVE_VPN_DNS_CHANGE_ALERT)) ==
+          chrome::MESSAGE_BOX_RESULT_YES);
+}
+
+bool BraveVpnDnsObserverService::IsDnsModeConfiguredByPolicy() const {
+  bool doh_policy_set =
+      policy_reader_ &&
+      !policy_reader_.Run(policy::key::kDnsOverHttpsMode).empty();
+  return doh_policy_set ||
+         SystemNetworkContextManager::GetStubResolverConfigReader()
+             ->ShouldDisableDohForManaged();
+}
+
+void BraveVpnDnsObserverService::ShowPolicyWarningMessage() {
+  if (!profile_prefs_->GetBoolean(prefs::kBraveVPNShowDNSPolicyWarningDialog)) {
+    return;
+  }
+
+  if (policy_callback_) {
+    std::move(policy_callback_).Run();
+    return;
+  }
+
+  chrome::ShowWarningMessageBoxWithCheckbox(
+      GetAnchorBrowserWindow(), l10n_util::GetStringUTF16(IDS_PRODUCT_NAME),
+      l10n_util::GetStringUTF16(IDS_BRAVE_VPN_DNS_POLICY_ALERT),
+      l10n_util::GetStringUTF16(IDS_BRAVE_VPN_DNS_POLICY_CHECKBOX),
+      base::BindOnce(&SkipDNSDialog, profile_prefs_));
+}
+
+void BraveVpnDnsObserverService::ShowMessageWhyWeOverrideDNSSettings() {
+  if (skip_notification_dialog_for_testing_)
+    return;
+  VpnNotificationDialogView::Show(chrome::FindLastActive());
+}
+
+void BraveVpnDnsObserverService::OnDNSPrefChanged() {
+  if (ignore_prefs_change_)
+    return;
+  bool is_dns_secure =
+      IsDNSSecure(SystemNetworkContextManager::GetStubResolverConfigReader()
+                      ->GetSecureDnsConfiguration(false));
+  if (is_dns_secure)
+    return;
+  // Reset saved config and keep user's choice.
+  if (!ShouldAllowExternalChanges()) {
+    ignore_prefs_change_ = true;
+    const auto& saved_dns_config =
+        profile_prefs_->GetDict(prefs::kBraveVPNUserConfig);
+    auto* servers_to_restore = saved_dns_config.FindString(kDohServersValue);
+    SetDNSOverHTTPSMode(SecureDnsConfig::kModeSecure,
+                        GetDoHServers(servers_to_restore));
+    ignore_prefs_change_ = false;
+  } else {
+    auto dns_config = SystemNetworkContextManager::GetStubResolverConfigReader()
+                          ->GetSecureDnsConfiguration(false);
+    SaveUserDNSConfig(profile_prefs_, dns_config);
+  }
+}
+
+void BraveVpnDnsObserverService::SetDNSOverHTTPSMode(
+    const std::string& mode,
+    const std::string& doh_providers) {
+  local_state_->SetString(::prefs::kDnsOverHttpsTemplates, doh_providers);
+  local_state_->SetString(::prefs::kDnsOverHttpsMode, mode);
+}
+
+void BraveVpnDnsObserverService::UnlockDNS() {
+  profile_prefs_->SetBoolean(prefs::kBraveVPNUserConfigLocked, false);
```

Comment:
this is what we discussed that we don't want to do and was the major thing to change from the original implementation. We should not be touching the actual user pref values, we should only be overriding by policy

Reply:
done